# 🧮 NumPy কী?

**NumPy** হলো Python-এর একটি মৌলিক লাইব্রেরি, যা **scientific computing (বৈজ্ঞানিক গণনা)**-এর জন্য ব্যবহৃত হয়।  
এটি এমন একটি লাইব্রেরি যা **multidimensional array object (একাধিক মাত্রার অ্যারে অবজেক্ট)**, কিছু **derived object** (যেমন masked array, matrix ইত্যাদি) এবং বিভিন্ন ধরনের **দ্রুতগতির অ্যারে অপারেশন** করার ফাংশন সরবরাহ করে।  

### ✳️ NumPy যা করতে পারে:
- Mathematical এবং Logical operation  
- Shape পরিবর্তন (Shape manipulation)  
- Sorting ও Selecting  
- Input/Output (I/O) অপারেশন  
- Discrete Fourier Transform  
- Basic Linear Algebra (যেমন: ম্যাট্রিক্সের যোগ-বিয়োগ-গুণ)  
- Statistical operations (গড়, variance ইত্যাদি)  
- Random simulation এবং আরও অনেক কিছু  

---

## 🧱 NumPy-এর মূল অংশ – `ndarray` অবজেক্ট

NumPy-এর সবচেয়ে গুরুত্বপূর্ণ অংশ হলো **`ndarray` object**,  
যা **n-dimensional homogeneous data array** ধারণ করে।  
এর মানে হলো, এটি একাধিক মাত্রার অ্যারে রাখে এবং সব উপাদান (elements) একই ডেটা টাইপের হয় (যেমন সবগুলো `int` বা সবগুলো `float`)।

---

## ⚖️ NumPy Arrays বনাম Python Sequences

| বিষয় | NumPy Array | Python List / Sequence |
|------|--------------|------------------------|
| **আকার (Size)** | নির্দিষ্ট (fixed) | পরিবর্তনযোগ্য (grow/shrink dynamically) |
| **ডেটা টাইপ** | সব উপাদান একই টাইপের | ভিন্ন ভিন্ন টাইপের হতে পারে |
| **পারফরম্যান্স** | খুব দ্রুত (C-এ লেখা) | তুলনামূলক ধীর |
| **অপারেশন** | গাণিতিক ও বৈজ্ঞানিক কাজে দক্ষ | তেমন দক্ষ নয় |

🧩 NumPy array তৈরি করার সময়ই এর আকার নির্ধারিত হয়।  
তুমি যদি আকার পরিবর্তন করো, তাহলে পুরনো অ্যারে ডিলিট হয়ে নতুন একটি তৈরি হয়।

---

## ⚙️ কেন NumPy ব্যবহার করা হয়?

NumPy ব্যবহার করলে বড় পরিসরের ডেটার ওপর জটিল গাণিতিক বা বৈজ্ঞানিক অপারেশনগুলো  
অনেক সহজ ও দ্রুত করা যায়।  
Python-এর সাধারণ `list` বা `tuple` দিয়ে করলে কোড বেশি লাগে এবং গতি কম হয়।

---

## 🌍 NumPy-এর গুরুত্ব

বেশিরভাগ বৈজ্ঞানিক ও গাণিতিক Python লাইব্রেরি (যেমন **Pandas**, **SciPy**, **TensorFlow**)  
সবই **NumPy array**-এর ওপর ভিত্তি করে তৈরি।  

তারা ইনপুট হিসেবে Python list নিতে পারে,  
কিন্তু প্রক্রিয়াকরণের আগে সেটাকে NumPy array-এ রূপান্তর করে নেয়।  
আউটপুট হিসেবেও সাধারণত NumPy array ফেরত দেয়।

---

## 💡 উদাহরণ


```

In [1]:

import numpy as np

# 1D array তৈরি
arr = np.array([1, 2, 3, 4, 5])
print("Array:", arr)

# কিছু অপারেশন
print("Mean:", np.mean(arr))
print("Standard Deviation:", np.std(arr))
print("Square:", np.square(arr))

Array: [1 2 3 4 5]
Mean: 3.0
Standard Deviation: 1.4142135623730951
Square: [ 1  4  9 16 25]



## 🔹 1️⃣ 1D Array (One-Dimensional Array)

> এক মাত্রিক অ্যারে হলো একটি সরল **list**-এর মতো —  
> যেখানে উপাদানগুলো একের পর এক থাকে, যেমন `[1, 2, 3, 4, 5]`।

🧠 ব্যাখ্যা:
- এখানে ndim = 1 মানে এটি এক-মাত্রিক অ্যারে।
- shape = (5,) মানে অ্যারেতে ৫টি উপাদান আছে, এক লাইনেই।


In [2]:

import numpy as np

# 1D array তৈরি
arr1 = np.array([10, 20, 30, 40, 50])
print("1D Array:\n", arr1)
print("Dimension:", arr1.ndim)
print("Shape:", arr1.shape)

1D Array:
 [10 20 30 40 50]
Dimension: 1
Shape: (5,)


## 2️⃣ 2D Array (Two-Dimensional Array)

- দুই মাত্রিক অ্যারে মানে একাধিক সারি (rows) ও কলাম (columns) —
- যেমন একটা ম্যাট্রিক্স বা টেবিল।

🧠 ব্যাখ্যা:

- ndim = 2 মানে এটি দুই মাত্রিক।
- shape = (3, 3) মানে ৩টি সারি ও ৩টি কলাম আছে।

In [3]:
# 2D array তৈরি ---can call it matrix
arr2 = np.array([[1,2,3],
                  [4,5,6],
                  [7,8,9]])
print('2D array:\n', arr2)
print('Dimention: ', arr2.ndim)
print('Shape: ', arr2.shape)

2D array:
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
Dimention:  2
Shape:  (3, 3)



## 🔹 3️⃣ 3D Array (Three-Dimensional Array)

- তিন মাত্রিক অ্যারে হলো অ্যারেগুলোর স্তূপ (stack) —
- অর্থাৎ একাধিক 2D ম্যাট্রিক্স একসাথে রাখা।
- একে সহজভাবে ভাবা যায়, যেন একাধিক পাতার নোটবুক 📘,
- প্রতিটি পাতাই একটি 2D array।

🧠 ব্যাখ্যা:

- এখানে ndim = 3 → তিন মাত্রিক।
- shape = (2, 2, 3) মানে:
- ২টা 2D ম্যাট্রিক্স (অর্থাৎ ২টি “layer”)
- প্রতিটি ম্যাট্রিক্সে ২টি সারি ও 3টি কলাম।

In [5]:
# 3D array তৈরি ---- can call it tensor
arr3 = np.array([[[1,2,3],
                [4,5,6]],
                [[7,8,9],
                [10,11,12]]])
print('3D array:\n', arr3)
print('Dimension: ', arr3.ndim)
print('Shape: ', arr3.shape)

3D array:
 [[[ 1  2  3]
  [ 4  5  6]]

 [[ 7  8  9]
  [10 11 12]]]
Dimension:  3
Shape:  (2, 2, 3)


## `dtype` (Data Type in NumPy)
- NumPy-তে dtype মানে Data Type।
- এটি বলে দেয় একটি NumPy array-এর প্রতিটি উপাদানের ডেটার ধরন কী — যেমন integer, float, string ইত্যাদি।

|         `dtype`        | মানে           | বর্ণনা                   |
| :--------------------: | :------------- | :----------------------- |
|   `int32` বা `int64`   | Integer        | পূর্ণ সংখ্যা             |
| `float32` বা `float64` | Floating point | দশমিকসহ সংখ্যা           |
|         `bool`         | Boolean        | True বা False মান        |
|        `complex`       | Complex Number | বাস্তব ও কাল্পনিক সংখ্যা |
|      `str` বা `<U`     | Unicode string | টেক্সট ডেটা              |


In [6]:
# dtype

a = np.array([1,2,3,4], dtype='int')
b = np.array([1,2,3,4,5], dtype='float')
c = np.array([1,2,3,4,5], dtype='complex')

print(a, a.dtype)
print(b, b.dtype)
print(c, c.dtype)

[1 2 3 4] int64
[1. 2. 3. 4. 5.] float64
[1.+0.j 2.+0.j 3.+0.j 4.+0.j 5.+0.j] complex128


## 🔢 np.arange() — (NumPy Range Function)

 - np.arange() হলো NumPy-এর একটি ফাংশন,
 - যা নির্দিষ্ট range (পরিসীমা) অনুযায়ী সমান ব্যবধানের সংখ্যা তৈরি করে।
- এক কথায় — এটি Python-এর range() ফাংশনের মতো,
- কিন্তু NumPy array আকারে আউটপুট দেয়

    ```
    np.arange(start, stop, step, dtype=None)
    ```
| Parameter | ব্যাখ্যা                             |
| :-------- | :----------------------------------- |
| `start`   | শুরু মান (ডিফল্ট 0)                  |
| `stop`    | শেষ মান (এই মানটি অন্তর্ভুক্ত হয় না) |
| `step`    | প্রতিটি মানের ব্যবধান (ডিফল্ট 1)     |
| `dtype`   | আউটপুটের ডেটাটাইপ (ঐচ্ছিক)           |


In [9]:
import numpy as np

# 🧾 উদাহরণ ১ — সাধারণ ব্যবহার:
arr = np.arange(10)  # 0 থেকে 9 পর্যন্ত সংখ্যা
print("Array:", arr)

# 🧾 উদাহরণ ২ — Start, Stop, Step সহ:
arr2 = np.arange(1, 10, 2)  # 1 থেকে 9 পর্যন্ত বিজোড় সংখ্যা
print("Array with step:", arr2)

# 🧾 উদাহরণ ৩ — Float মান সহ:
arr3 = np.arange(0.5, 5.5, 1.0)  # 0.5 থেকে 5.5 পর্যন্ত 1.0 করে বৃদ্ধি
print("Float array:", arr3)

# ⚙️ dtype নির্দিষ্ট করা:

arr4 = np.arange(5, 10, 2, dtype='float') # 5 থেকে 10 পর্যন্ত 3 করে বৃদ্ধি, float টাইপ
print("Array with specific dtype:", arr4)

Array: [0 1 2 3 4 5 6 7 8 9]
Array with step: [1 3 5 7 9]
Float array: [0.5 1.5 2.5 3.5 4.5]
Array with specific dtype: [5. 7. 9.]


### সংক্ষেপে মনে রাখো:
✅ `np.arange(start, stop, step)`
- Start থেকে Stop পর্যন্ত Step ব্যবধানে সংখ্যা তৈরি করে
- Stop মানটি অন্তর্ভুক্ত হয় না
- আউটপুট দেয় NumPy array

## 🔄 reshape() — (Array-এর আকার পরিবর্তন)
`reshape()` ফাংশন ব্যবহার করে আমরা কোনো NumPy array-এর `shape বা dimension` পরিবর্তন করতে পারি,
- কিন্তু ডেটা অপরিবর্তিত থাকে।
- অর্থাৎ — একই ডেটা, নতুন আকারে সাজানো হয়।
```
array.reshape(new_shape)
```

In [10]:
import numpy as np

# 🧾 উদাহরণ ১ — 1D → 2D
arr = np.arange(6) # 0 থেকে 5 পর্যন্ত সংখ্যা
print("Original 1D array:", arr)

new_arr = arr.reshape(2, 3) # 2 rows এবং 3 columns
print('Reshaped to 2D array:\n', new_arr)

Original 1D array: [0 1 2 3 4 5]
Reshaped to 2D array:
 [[0 1 2]
 [3 4 5]]


In [11]:
# 🧾 উদাহরণ ২ — 1D → 3D
arr = np.arange(8) # 0 থেকে 7 পর্যন্ত সংখ্যা
print("Original 1D array:", arr)

reshaped = arr.reshape(2, 2, 2) # 2 blocks, 2 rows, 2 columns
print('Reshaped to 3D array:\n', reshaped)

Original 1D array: [0 1 2 3 4 5 6 7]
Reshaped to 3D array:
 [[[0 1]
  [2 3]]

 [[4 5]
  [6 7]]]


In [12]:
# 🧾 উদাহরণ ৩ — স্বয়ংক্রিয় dimension (Auto detect with -1)

arr = np.arange(12) # 0 থেকে 11 পর্যন্ত সংখ্যা
print("Original 1D array:", arr)

reshaped_auto = arr.reshape(3, -1) # 3 rows এবং auto-detect column
print('Reshaped with auto-detect:\n', reshaped_auto)

Original 1D array: [ 0  1  2  3  4  5  6  7  8  9 10 11]
Reshaped with auto-detect:
 [[ 0  1  2  3]
 [ 4  5  6  7]
 [ 8  9 10 11]]


🔹 **এখানে `-1` মানে “NumPy নিজে বুঝে নেবে কত কলাম দরকার”।**

### ⚠️ গুরুত্বপূর্ণ নিয়ম:

- ✅ মোট উপাদানের সংখ্যা (elements) একই থাকতে হবে
- যেমন, arr-এর দৈর্ঘ্য যদি 8 হয়,
- তাহলে (2, 4), (4, 2), (2, 2, 2) — এসব বৈধ,
- কিন্তু (3, 3) অবৈধ কারণ 9 ≠ 8।

## 🧩 .reshape() ও .shape এর পার্থক্য
| বিষয়        | `.reshape()`               | `.shape`              |
| :---------- | :------------------------- | :-------------------- |
| কাজ         | Array-এর আকার পরিবর্তন করে | বর্তমান আকার জানায়    |
| আউটপুট      | নতুন array রিটার্ন করে     | Tuple আকারে shape দেয় |
| পরিবর্তন হয় | না, নতুন কপি তৈরি হয়       | না, শুধু তথ্য দেয়     |


In [ ]:
# examples with reshape and shape
arr = np.arange(1, 11).reshape(2, 5) # 1 থেকে 10 পর্যন্ত সংখ্যা, 2 rows এবং 5 columns
print("Reshaped array:\n", arr) 
print("Shape:", arr.shape) 

Reshaped array:
 [[ 1  2  3  4  5]
 [ 6  7  8  9 10]]
Shape: (2, 5)


##  `np.zeros()` — শূন্য দিয়ে পূর্ণ array তৈরি
- np.zeros(shape, dtype=float) ব্যবহার করে আমরা সব উপাদান শূন্য (0) দিয়ে পূর্ণ একটি NumPy array তৈরি করতে পারি।
- shape → array-এর আকার (1D, 2D, 3D ইত্যাদি)
- dtype → ডেটার ধরন (default: float)

In [6]:
# 🧾 উদাহরণ ১ — 1D Array

import numpy as np

arr = np.zeros(5) # 5 টি শূন্য মান সহ 1D array
print("1D Array of zeros:", arr)
print("Shape: ", arr.shape)
print("Dtype: ", arr.dtype)
print('Dimension: ', arr.ndim)

1D Array of zeros: [0. 0. 0. 0. 0.]
Shape:  (5,)
Dtype:  float64
Dimension:  1


In [7]:
# 🧾 উদাহরণ ২ — 2D Array

arr = np.zeros((2, 3),  dtype=np.float64) # 2 rows এবং 3 columns সহ 2D array
print("2D Array of zeros:", arr)
print("Shape: ", arr.shape)
print("Dtype: ", arr.dtype)
print("Dimension: ", arr.ndim)

2D Array of zeros: [[0. 0. 0.]
 [0. 0. 0.]]
Shape:  (2, 3)
Dtype:  float64
Dimension:  2


## `np.ones()` — এক দিয়ে পূর্ণ array তৈরি
- np.ones(shape, dtype=float) ব্যবহার করে আমরা সব উপাদান এক (1) দিয়ে পূর্ণ একটি NumPy array তৈরি করতে পারি।
- shape → array-এর আকার
- dtype → ডেটার ধরন (default: float)

In [14]:
# 🧾 উদাহরণ ১ — 1D Array

arr = np.ones(5, dtype=np.float64) # 5 টি এক মান সহ 1D array
print('1D Array of ones: ', arr)
print("Shape: ", arr.shape)
print("Dtype: ", arr.dtype)
print('Dimension: ', arr.ndim)

1D Array of ones:  [1. 1. 1. 1. 1.]
Shape:  (5,)
Dtype:  float64
Dimension:  1


In [9]:
# 🧾 উদাহরণ ২ — 2D Array

arr = np.ones((3, 4)) # 3 rows এবং 4 columns সহ 2D array
print('2D Array of ones:\n', arr)
print("Shape: ", arr.shape)
print("Dtype: ", arr.dtype)
print("Dimension: ", arr.ndim)

2D Array of ones:
 [[1. 1. 1. 1.]
 [1. 1. 1. 1.]
 [1. 1. 1. 1.]]
Shape:  (3, 4)
Dtype:  float64
Dimension:  2


## 🧩 মূল পার্থক্য
| Function     | Output      | ডিফল্ট dtype |
| ------------ | ----------- | ------------ |
| `np.zeros()` | সব উপাদান 0 | float64      |
| `np.ones()`  | সব উপাদান 1 | float64      |


In [15]:
# examples of 0's & 1's

np.zeros(10).reshape(5, 2) # 10 টি শূন্য মান সহ 5 rows এবং 2 columns সহ 2D array

array([[0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.]])

In [18]:
np.ones(10).reshape(2, 5) # 12 টি এক মান সহ 2 rows এবং 5 columns সহ 2D array

array([[1., 1., 1., 1., 1.],
       [1., 1., 1., 1., 1.]])

## 🎲 np.random — Random Number Generation
📘 কী?

np.random হল NumPy-এর একটি সাবমডিউল, যা random numbers বা random arrays তৈরি করতে ব্যবহৃত হয়।
এটি বিভিন্ন distribution অনুযায়ী সংখ্যা জেনারেট করতে পারে, যেমন: uniform, normal, integer, choice ইত্যাদি।

🔹 সাধারণ Syntax:    `np.random.function_name(parameters)`


### function_name এর মধ্যে থাকতে পারে:
 - `rand()` → Uniform distribution [0,1)
 - `randint()` → নির্দিষ্ট range-এর মধ্যে integer
 - `randn()` → Standard Normal distribution
 - `choice()` → নির্দিষ্ট array থেকে random selection
 
 

In [32]:
# 🧾 উদাহরণ ১ — Uniform Random Number

import numpy as np
arr = np.random.rand(5) # 0 থেকে 1 পর্যন্ত 5 টি র্যান্ডম সংখ্যা
print("Uniform random numbers between 0 and 1:", arr)

Uniform random numbers between 0 and 1: [0.58675117 0.96525531 0.60703425 0.27599918 0.29627351]


In [21]:
# 🧾 উদাহরণ ২ — Random Integers

arr = np.random.randint(1, 10, 5) # 1 থেকে 10 পর্যন্ত random integer 5টি
print("Random integers between 1 and 10:", arr )

Random integers between 1 and 10: [9 7 7 4 6]


In [22]:
# 🧾 উদাহরণ ৩ — Standard Normal Distribution

arr = np.random.randn(5) # mean=0 এবং std=1 সহ 5 টি র্যান্ডম সংখ্যা
print("Random numbers from standard normal distribution:", arr)

Random numbers from standard normal distribution: [-0.67151176 -0.81289465 -0.31000543  0.48451351  0.44414756]


In [25]:
# 🧾 উদাহরণ ৪ — Random Choice
# একটি array থেকে random selection

arr = np.random.choice([10, 20, 30, 40, 50, 60, 70, 80, 90, 100], size=5) # প্রদত্ত তালিকা থেকে 5 টি এলোমেলো মান
print("Random choice from list:", arr)

Random choice from list: [60 90 30 40 90]


In [29]:
# ✅ Reproducibility: যদি একই random numbers বারবার চাই, np.random.seed(seed_number) ব্যবহার করো।

np.random.seed(42) # যেকোনো একটি সংখ্যা দিয়ে seed সেট করা যায়
np.random.randint(1, 50, 10)

array([39, 29, 15, 43,  8, 21, 39, 19, 23, 11], dtype=int32)

In [30]:
np.random.randint(1, 50, 10)

array([11, 24, 36, 40, 24,  3, 22,  2, 24, 44], dtype=int32)


 # 🎲 NumPy Random Functions – Summary

NumPy-এর `np.random` মডিউল দিয়ে বিভিন্ন ধরনের random numbers বা arrays তৈরি করা যায়।  
নিচের টেবিলটি মূল ফাংশনগুলো সংক্ষেপে দেখাচ্ছে:

| Function | কাজ | Syntax | উদাহরণ |
|----------|-----|--------|--------|
| `rand(d0, d1, …, dn)` | Uniform distribution [0,1) অনুযায়ী random float | `np.random.rand(5)` | `[0.5488, 0.7151, 0.6027, 0.5448, 0.4236]` |
| `randn(d0, d1, …, dn)` | Standard Normal distribution (mean=0, std=1) | `np.random.randn(5)` | `[0.7692, -0.3556, 1.4434, 0.0530, -1.3776]` |
| `randint(low, high, size)` | নির্দিষ্ট range-এর মধ্যে random integer | `np.random.randint(0, 10, 5)` | `[3, 7, 2, 9, 1]` |
| `choice(array, size, replace=True, p=None)` | নির্দিষ্ট array থেকে random selection | `np.random.choice([10,20,30], size=3)` | `[20,10,30]` |
| `seed(seed_number)` | Random numbers reproducible করা | `np.random.seed(42)` | একই seed দিলে একই random numbers বারবার পাওয়া যায় |

---

### 🔹 উদাহরণ কোড (সব একসাথে):

```python
import numpy as np

# Seed set করলে reproducible হয়
np.random.seed(42)

# 1. Uniform [0,1)
print("rand:", np.random.rand(5))

# 2. Standard Normal
print("randn:", np.random.randn(5))

# 3. Random integers
print("randint:", np.random.randint(0, 10, 5))

# 4. Random choice
print("choice:", np.random.choice([10, 20, 30, 40], size=3))


In [33]:
import numpy as np

# Seed set করলে reproducible হয়
np.random.seed(42)

# 1. Uniform [0,1)
print("rand:", np.random.rand(5))

# 2. Standard Normal
print("randn:", np.random.randn(5))

# 3. Random integers
print("randint:", np.random.randint(0, 10, 5))

# 4. Random choice
print("choice:", np.random.choice([10, 20, 30, 40], size=3))

rand: [0.37454012 0.95071431 0.73199394 0.59865848 0.15601864]
randn: [ 0.27904129  1.01051528 -0.58087813 -0.52516981 -0.57138017]
randint: [5 8 0 9 2]
choice: [40 30 40]


## 📏 np.linspace() — Linearly spaced numbers
#### ফাংশনটি start থেকে stop পর্যন্ত সমান ব্যবধানের (linearly spaced) সংখ্যার array তৈরি করে।
- np.linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None) 

- start → array-এর প্রথম মান
- stop → array-এর শেষ মান
- num → কতগুলো value হবে (default=50)
- endpoint → True হলে stop অন্তর্ভুক্ত হবে, False হলে হবে না
- retstep → True দিলে step size আউটপুট হবে
- dtype → ডেটার ধরন

In [34]:
# 🧾 উদাহরণ ১ — সাধারণ ব্যবহার

import numpy as np
np.linspace(10, 50, 5) # 10 থেকে 50 পর্যন্ত 5 টি সমানভাবে spaced সংখ্যা
# lower range -> upper range-> number of items you want to generate


array([10., 20., 30., 40., 50.])

- ব্যাখ্যা: 10 থেকে 50 পর্যন্ত 5টি সমান ব্যবধানের সংখ্যা তৈরি হয়েছে।

In [35]:
# 🧾 উদাহরণ ২ — Stop value বাদ দিয়ে
np.linspace(10, 50, 5, endpoint=False) # 10 থেকে 50 পর্যন্ত 5 টি সমানভাবে spaced সংখ্যা, কিন্তু 50 অন্তর্ভুক্ত নয়

array([10., 18., 26., 34., 42.])

- Stop (50) অন্তর্ভুক্ত হয়নি, spacing auto-adjusted।

In [36]:
# 🧾 উদাহরণ ৩ — Step size দেখতে

np.linspace(10, 50, 5, retstep=True) # 10 থেকে 50 পর্যন্ত 5 টি সমানভাবে spaced সংখ্যা এবং step size দেখাও

(array([10., 20., 30., 40., 50.]), np.float64(10.0))

### ⚙️ arange বনাম linspace
| Function                     | মূল পার্থক্য                                     | উদাহরণ                                            |
| ---------------------------- | ------------------------------------------------ | ------------------------------------------------- |
| `arange(start, stop, step)`  | step নির্দিষ্ট, stop মান অন্তর্ভুক্ত না হতে পারে | `np.arange(0, 1, 0.2)` → `[0. ,0.2,0.4,0.6,0.8]`  |
| `linspace(start, stop, num)` | num নির্দিষ্ট, step auto-calculated              | `np.linspace(0, 1, 5)` → `[0. ,0.25,0.5,0.75,1.]` |

### সংক্ষেপে মনে রাখো:

- linspace → সমান spacing সহ নির্দিষ্ট সংখ্যক value
- arange → নির্দিষ্ট step size অনুযায়ী value
- Control করতে num, endpoint, retstep ব্যবহার করো


###  Practice:  
- `np.linspace() দিয়ে 1D → 2D → 3D array তৈরি করা।`

In [37]:
# 1️⃣ 1D Array

import numpy as np
arr = np.linspace(10, 50, 5) # 10 থেকে 50 পর্যন্ত 5টি সমান ব্যবধানের সংখ্যা 
print('1D Array:\n', arr)
print('Dimension:', arr.ndim)
print('Shape:', arr.shape)

1D Array:
 [10. 20. 30. 40. 50.]
Dimension: 1
Shape: (5,)


In [38]:
# 2️⃣ 2D Array (reshape() ব্যবহার করে)

arr = np.linspace(1, 10, 6).reshape(2, 3) # 1 থেকে 10 পর্যন্ত 6টি সমান ব্যবধানের সংখ্যা 
print('2D Array:\n', arr)
print('Dimention: ', arr.ndim)
print('Shape: ', arr.shape)


2D Array:
 [[ 1.   2.8  4.6]
 [ 6.4  8.2 10. ]]
Dimention:  2
Shape:  (2, 3)


- ব্যাখ্যা: `reshape(2,3)` → 2 rows, 3 columns

In [39]:
# 3️⃣ 3D Array (reshape() ব্যবহার করে)

arr = np.linspace(1, 24, 24).reshape(2, 3, 4) # 1 থেকে 24 পর্যন্ত 24টি সমান ব্যবধানের সংখ্যা 
print('3D Array:\n', arr)
print('Dimension: ', arr.ndim)
print('Shape: ', arr.shape)


3D Array:
 [[[ 1.  2.  3.  4.]
  [ 5.  6.  7.  8.]
  [ 9. 10. 11. 12.]]

 [[13. 14. 15. 16.]
  [17. 18. 19. 20.]
  [21. 22. 23. 24.]]]
Dimension:  3
Shape:  (2, 3, 4)


- ব্যাখ্যা: `reshape(2,3,4)` → 2 layers, প্রতিটি layer 3 rows × 4 columns

### ✨ সারাংশ:
- 1D → সরল লাইন
- 2D → ম্যাট্রিক্স / টেবিল
- 3D → একাধিক ম্যাট্রিক্সের stack
- `np.linspace()` দিয়ে আমরা সমান spacing সহ সংখ্যা তৈরি করি, `reshape()` দিয়ে আকার পরিবর্তন করি

## `np.identity()` — Identity Matrix

- `np.identity(n, dtype=float)` একটি `n×n identity matrix` তৈরি করে।
- `Identity matrix` হলো `square matrix`, যার মূল `diagonal (main diagonal)` এ `1` থাকে, এবং বাকি সব element `0`।
- ইঞ্জিনিয়ারিং, লিনিয়ার অ্যালজেব্রা, ও ম্যাট্রিক্স অপারেশনে খুব গুরুত্বপূর্ণ।

In [49]:
# 🧾 উদাহরণ ১ — 3×3 Identity Matrix
# identity matrix হল একটি স্কয়ার matrix যার প্রধান ডায়াগোনালে 1 থাকে এবং বাকি সব এলিমেন্ট 0 থাকে


import numpy as np
# arr = np.identity(3) # 3x3 identity matrix
arr = np.eye(3) # 3×3 identity matrix
print("3x3 Identity Matrix:\n", arr)

3x3 Identity Matrix:
 [[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]


In [50]:
# 🧾 উদাহরণ ২ — ডেটাটাইপ পরিবর্তন

arr = np.eye(5, dtype='int') # 5×5 identity matrix, integer type
print("5x5 Identity Matrix with integer type:\n", arr)

5x5 Identity Matrix with integer type:
 [[1 0 0 0 0]
 [0 1 0 0 0]
 [0 0 1 0 0]
 [0 0 0 1 0]
 [0 0 0 0 1]]


## ⚙️ তুলনা — `zeros, ones, identity`
| Function          | Output                 | মূল ব্যবহার                           |
| ----------------- | ---------------------- | ------------------------------------- |
| `np.zeros((n,m))` | সব 0                   | Initialization                        |
| `np.ones((n,m))`  | সব 1                   | Initialization                        |
| `np.identity(n) or np.eye(n)`  | মূল diagonal 1, বাকি 0 | লিনিয়ার অ্যালজেব্রা, identity matrix |

### 💡 সংক্ষেপে মনে রাখো:
- ✅ Identity matrix হলো square matrix যেখানে main diagonal = 1, বাকি সব = 0
- ✅ np.identity(n) → n×n identity matrix তৈরি করে
- ✅ dtype দ্বারা integer বা float নির্ধারণ করা যায়
- ✅ Linear algebra, matrix multiplication ইত্যাদিতে গুরুত্বপূর্ণ

# `np.identity()` vs `np.eye()`

| বৈশিষ্ট্য | `np.identity()` | `np.eye()` |
|------------|-----------------|------------|
| কাজ | Identity matrix তৈরি করে | Identity matrix বা custom diagonal matrix তৈরি করে |
| Square matrix (row = column) | ✅ শুধুমাত্র square matrix | ✅ করতে পারে |
| Rectangular matrix (row ≠ column) | ❌ সম্ভব না | ✅ সম্ভব |
| Diagonal shift (`k`) | ❌ নেই | ✅ আছে |
| Flexibility | কম | বেশি |
| ব্যবহার | সাধারণ identity matrix | Advanced/custom matrix তৈরিতে |

## উদাহরণ

### Example: `np.identity()`

```python
import numpy as np
np.identity(3)
```

Output:

```python
array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])
```

### Example: `np.eye()`

```python
np.eye(3,4)
```

Output:

```python
array([[1., 0., 0., 0.],
       [0., 1., 0., 0.],
       [0., 0., 1., 0.]])
```

### Example: `np.eye()` with shifted diagonal

```python
np.eye(4, k=1)
```

Output:

```python
array([[0., 1., 0., 0.],
       [0., 0., 1., 0.],
       [0., 0., 0., 1.],
       [0., 0., 0., 0.]])
```

# 🟢 NumPy Array Creation Functions – Quick Reference

NumPy array তৈরি করার বিভিন্ন গুরুত্বপূর্ণ ফাংশন সংক্ষেপে:

| Function | কাজ | Syntax | উদাহরণ |
|----------|-----|--------|--------|
| `np.arange()` | নির্দিষ্ট step অনুযায়ী সমান ব্যবধানের সংখ্যা | `np.arange(start, stop, step)` | `np.arange(0, 10, 2)` → `[0 2 4 6 8]` |
| `np.linspace()` | নির্দিষ্ট সংখ্যক সমান spaced numbers | `np.linspace(start, stop, num)` | `np.linspace(0, 10, 5)` → `[0. ,2.5,5.,7.5,10.]` |
| `np.zeros()` | সব উপাদান 0 | `np.zeros(shape, dtype)` | `np.zeros((2,3))` → `[[0. 0. 0.],[0. 0. 0.]]` |
| `np.ones()` | সব উপাদান 1 | `np.ones(shape, dtype)` | `np.ones((3,2), dtype=int)` → `[[1 1],[1 1],[1 1]]` |
| `np.identity()` | n×n identity matrix | `np.identity(n, dtype)` | `np.identity(3)` → `[[1. 0. 0.],[0. 1. 0.],[0. 0. 1.]]` |
| `np.eye()` | m×n diagonal matrix | `np.eye(N, M=None, k=0, dtype=float)` | `np.eye(3, k=1)` → `[[0. 1. 0.],[0. 0. 1.],[0. 0. 0.]]` |
| `np.random.rand()` | Uniform [0,1) distribution float array | `np.random.rand(d0, d1, …)` | `np.random.rand(2,3)` → `[[0.3745 0.9507 0.7319],[0.5986 0.1560 0.1559]]` |
| `np.random.randn()` | Standard normal distribution | `np.random.randn(d0,d1,…)` | `np.random.randn(3)` → `[0.4967 -0.1382 0.6476]` |
| `np.random.randint()` | Random integer array | `np.random.randint(low, high, size)` | `np.random.randint(0,10,5)` → `[6 3 7 4 6]` |
| `np.random.choice()` | Array থেকে random selection | `np.random.choice(array, size)` | `np.random.choice([10,20,30],3)` → `[20 10 30]` |

---

## 🔹 উদাহরণ কোড (Compact)

```python
import numpy as np

# 1D Arrays
a1 = np.arange(0,10,2)
a2 = np.linspace(0,10,5)
a3 = np.zeros(3)
a4 = np.ones(4)
a5 = np.identity(3)
a5 = np.eye(6)

# Random
np.random.seed(42)
r1 = np.random.rand(3)
r2 = np.random.randn(3)
r3 = np.random.randint(0,10,5)
r4 = np.random.choice([10,20,30],3)

# Print
print("arange:", a1)
print("linspace:", a2)
print("zeros:", a3)
print("ones:", a4)
print("identity:\n", a5)
print("eye:\n", a6)
print("rand:", r1)
print("randn:", r2)
print("randint:", r3)
print("choice:", r4)


In [52]:
import numpy as np

# 1D Arrays
a1 = np.arange(0,10,2)
a2 = np.linspace(0,10,5)
a3 = np.zeros(3)
a4 = np.ones(4)
a5 = np.identity(3)
a6 = np.eye(4)

# Random
np.random.seed(42)
r1 = np.random.rand(3)
r2 = np.random.randn(3)
r3 = np.random.randint(0,10,5)
r4 = np.random.choice([10,20,30],3)

# Print
print("arange:", a1)
print("linspace:", a2)
print("zeros:", a3)
print("ones:", a4)
print("identity:\n", a5)
print("eye:\n", a6)
print("rand:", r1)
print("randn:", r2)
print("randint:", r3)
print("choice:", r4)

arange: [0 2 4 6 8]
linspace: [ 0.   2.5  5.   7.5 10. ]
zeros: [0. 0. 0.]
ones: [1. 1. 1. 1.]
identity:
 [[1. 0. 0.]
 [0. 1. 0.]
 [0. 0. 1.]]
eye:
 [[1. 0. 0. 0.]
 [0. 1. 0. 0.]
 [0. 0. 1. 0.]
 [0. 0. 0. 1.]]
rand: [0.37454012 0.95071431 0.73199394]
randn: [-1.11188012  0.31890218  0.27904129]
randint: [7 2 5 4 1]
choice: [20 20 20]
